Assets/Mirror/Editor/EnterPlayModeSettingsCheck.cs

```diff
@@ -9,28 +9,56 @@ namespace Mirror
 {
     public class EnterPlayModeSettingsCheck : MonoBehaviour
     {
-#if UNITY_2019_3_OR_NEWER
         [InitializeOnLoadMethod]
         static void OnInitializeOnLoad()
         {
-            // check immediately on load
+#if UNITY_2019_3_OR_NEWER
+            // We can't support experimental "Enter Play Mode Options" mode
+            // Check immediately on load, and before entering play mode, and warn the user
             CheckPlayModeOptions();
+#endif
 
-            // check each time we press play. OnLoad is only called once and
-            // wouldn't detect editor-time setting changes
+            // Hook this event to see if we have a good weave every time
+            // user attempts to enter play mode or tries to do a build
             EditorApplication.playModeStateChanged += OnPlayModeStateChanged;
         }
 
         static void OnPlayModeStateChanged(PlayModeStateChange state)
         {
-            // only check when entering play mode. no need to show it again
-            // when exiting.
-            if (state == PlayModeStateChange.EnteredPlayMode)
+            // Per Unity docs, this fires "when exiting edit mode before the Editor is in play mode".
+            // This doesn't fire when closing the editor.
+            if (state == PlayModeStateChange.ExitingEditMode)
             {
+                CheckSuccessfulWeave();
+
+#if UNITY_2019_3_OR_NEWER
+                // We can't support experimental "Enter Play Mode Options" mode
+                // Check and prevent entering play mode if enabled
                 CheckPlayModeOptions();
+#endif
             }
         }
 
+        static void CheckSuccessfulWeave()
+        {
+            // Check if last weave result was successful
+            if (!SessionState.GetBool("MIRROR_WEAVED", false))
+            {
+                // Last weave result was a failure...try to weave again
+                // Faults will show in the console that may have been cleared by "Clear on Play"
+                Weaver.CompilationFinishedHook.WeaveExistingAssemblies();
+
+                // Did that clear things up for us?
+                if (!SessionState.GetBool("MIRROR_WEAVED", false))
```

Comment:
The original intent of MIRROR_WEAVED session state was to solve one problem:

If some assemblies got built before the weaver,  then the weaver would not weave them. 
So the very first time the weaver is loaded we rebuilt assemblies.   That way it is guaranteed that assemblies are always built after the weaver.   

After the weaver has been set up the first time.  We don't want to reload assemblies anymore,  because the weaver is already hooked to assemblyCompilationFinished

But this is using it for something else.

Reply:
So extending the usage of MIRROR_WEAVED is ill-advised the way I'm doing it?  Should I introduce another session bool instead?

Reply:
I am not sure yet,  let me chew on it first for a few hours.

Reply:
If there are any weaver errors it goes false until weave is called again.

Reply:
I've done some work on this, put in a new session bool, reinstated the RequestScriptReload, and put the event wire up back to the top.  The only gotcha I've got now is after fixing the issue the weaver was complaining about, I get a good weave, but it insists on weaving again before entering play mode for the first time after that.  Subsequent plays are fine with no more extra weaves, even if I change something without making a weaver error.